From edb9b2986fe912ba615666f24a0dc248aa0ca5b4 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Vesa=20J=C3=A4=C3=A4skel=C3=A4inen?=
 <vesa.jaaskelainen@vaisala.com>
Date: Sun, 30 Jul 2023 08:08:15 +0300
Subject: [PATCH] ta: pkcs11: on RSA public key import calculate
 CKA_MODULUS_BITS
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

When RSA public key is created with C_CreateObject() CKA_MODULUS_BITS is
not allowed to be specified.

The tooling like pkcs11-tool expects that the value is there.

In specification it is not specified clearly that it needs to be
calculated but it is assumed to be in the RSA public key object.

Calculate the value and add it to the object during RSA public key import.

Signed-off-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>
---
 ta/pkcs11/src/pkcs11_attributes.c | 51 +++++++++++++++++++++++++++++++
 1 file changed, 51 insertions(+)

diff --git a/ta/pkcs11/src/pkcs11_attributes.c b/ta/pkcs11/src/pkcs11_attributes.c
index 2bdfd7ed404..eb2173e402c 100644
--- a/ta/pkcs11/src/pkcs11_attributes.c
+++ b/ta/pkcs11/src/pkcs11_attributes.c
@@ -794,6 +794,53 @@ static enum pkcs11_rc create_pub_key_attributes(struct obj_attrs **out,
 	return set_attributes_opt_or_null(out, temp, oon, oon_count);
 }
 
+static enum pkcs11_rc
+create_pub_key_rsa_generated_attributes(struct obj_attrs **out,
+					struct obj_attrs *temp,
+					enum processing_func function)
+{
+	uint32_t key_bits = 0;
+	void *a_ptr = NULL;
+	uint32_t a_size = 0;
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+
+	if (function != PKCS11_FUNCTION_IMPORT)
+		return rc;
+
+	/* Calculate CKA_MODULUS_BITS */
+
+	if (get_attribute_ptr(temp, PKCS11_CKA_MODULUS,
+			      &a_ptr, &a_size) || !a_ptr) {
+		EMSG("No CKA_MODULUS attribute found in public key");
+		return PKCS11_CKR_ATTRIBUTE_TYPE_INVALID;
+	}
+
+	key_bits = a_size * 8;
+
+	return add_attribute(out, PKCS11_CKA_MODULUS_BITS, &key_bits,
+			     sizeof(key_bits));
+}
+
+static enum pkcs11_rc
+create_pub_key_generated_attributes(struct obj_attrs **out,
+				    struct obj_attrs *temp,
+				    enum processing_func function)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+
+	switch (get_key_type(*out)) {
+	case PKCS11_CKK_RSA:
+		rc = create_pub_key_rsa_generated_attributes(out, temp,
+							     function);
+		break;
+	default:
+		/* no-op */
+		break;
+	}
+
+	return rc;
+}
+
 static enum pkcs11_rc create_priv_key_attributes(struct obj_attrs **out,
 						 struct obj_attrs *temp)
 {
@@ -1122,6 +1169,10 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
 		break;
 	case PKCS11_CKO_PUBLIC_KEY:
 		rc = create_pub_key_attributes(&attrs, temp, function);
+		if (rc)
+			goto out;
+		rc = create_pub_key_generated_attributes(&attrs, temp,
+							 function);
 		break;
 	case PKCS11_CKO_PRIVATE_KEY:
 		rc = create_priv_key_attributes(&attrs, temp);
